Support cartesian shape with doc values

craigtaverner opened the pull request:
Also refactoring testing infrastructure to make testing easier to share across different test classes.

craigtaverner committed:
Support shape with doc values

Also refactoring testing infrastructure to make testing easier to share
across different test classes.

elasticmachine commented:
Pinging @elastic/es-analytics-geo (Team:Analytics)

elasticsearchmachine commented:
Hi @craigtaverner, I've created a changelog YAML for you.

craigtaverner committed:
Update docs/changelog/88487.yaml

iverase commented on x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/index/mapper/ShapeDocValuesQuery.java:
```diff
@@ -0,0 +1,197 @@
+
+    protected abstract Component2D create(GEOMETRY geometry);
+
+    protected abstract Component2D create(GEOMETRY[] geometries);
```
this is nice, do we need two methods? would not be enough with something like:
```
protected abstract Component2D create(GEOMETRY... geometry);
```

craigtaverner replied:
That is what I did originally, but it turns out varargs does not play with generics, so in the generic class you have to use the two methods to get the same effect.

craigtaverner replied:
I think it is because with type erasure we cannot tell the difference between Object[] and Object (since Object[] is also an Object).

iverase commented on x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/index/mapper/ShapeDocValuesQuery.java:
```diff
@@ -0,0 +1,197 @@
+
+    /** Override if special cases, like dateline support, should be considered */
+    protected boolean addSpecialCase(List<Component2D> components2D, GEOMETRY geometry) {
+        return false;
```
Not strong opinion but I wonder if we should change the API to somethng like:

```
    protected abstract Component2D create(GEOMETRY... geometry);

    protected abstract void add(List<Component2D> components2D, GEOMETRY geometry);
```

So we don't need this special case.

craigtaverner replied:
Interesting approach. Is easier to read and reason about, so I've changed it to that, sort-of.

iverase commented on x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/index/fielddata/LabelPositionVisitor.java:
```diff
@@ -7,54 +7,58 @@
  * TODO: We could instead choose the point closer to the centroid which improves unbalanced trees
  */
-public class LabelPositionVisitor implements TriangleTreeReader.Visitor {
+public class LabelPositionVisitor<T extends ToXContentFragment> implements TriangleTreeReader.Visitor {
```
Do we need to extend ToXContentFragment here? I think we don't need any method from that interface.

iverase commented on server/src/main/java/org/elasticsearch/index/mapper/ShapeIndexer.java:
```diff
@@ -0,0 +1,18 @@
+
+import java.util.List;
+
+public interface ShapeIndexer {
```
Add some java docs?

iverase commented on x-pack/plugin/spatial/src/test/java/org/elasticsearch/xpack/spatial/index/mapper/ShapeFieldMapperTests.java:
```diff
@@ -72,6 +72,12 @@ protected void registerParameters(ParameterChecker checker) throws IOException {
         });
     }
 
+    @Override
```
I am missing here explicit test for doc_values, like that they are generated when set the flag to true or that the default value is true in current version but. false in previous versions.

craigtaverner replied:
Added in 1066e01c98b3d8f5bd9873803366d5a2d7785f80

iverase commented on x-pack/plugin/spatial/src/test/java/org/elasticsearch/xpack/spatial/index/mapper/ShapeFieldMapperTests.java:
```diff
@@ -72,6 +72,12 @@ protected void registerParameters(ParameterChecker checker) throws IOException {
+    @Override
+    protected boolean supportsSearchLookup() {
+        // TODO: is this really true? We have failing tests, but they look like test issues not core issues
+        return false;
```
It seems that this was not needed before because we did not have doc values so it was equivalent to set this to false. Now we need to do it explicitly.

iverase commented on x-pack/plugin/spatial/src/main/java/org/elasticsearch/xpack/spatial/index/fielddata/ShapeValues.java:
```diff
@@ -0,0 +1,259 @@
+ *
+ * There is just one value for one document.
+ */
+public abstract class ShapeValues<T extends ToXContentFragment> {
```
Do we need this class in this PR?

craigtaverner replied:
Done in e52ba263c3d118caf23d1da4358b9f8d95076702

craigtaverner replied:
Good catch, indeed. I originally pulled in some of the other generic refactoring from the prototype, realised I did not need it and reverted the classes that extend this class, but forgot to remove this class itself!

iverase reviewed (changes requested):
This is looking pretty solid, just left some small comments. 

The most important thing to improve is the test coverage of the doc_value flag in ShapeFieldMapperTests, in particular making sure that set the flag to true, it generates doc_values and that the default value changes depending on the versions.

craigtaverner committed:
Fixed test regression with dateline wrapping

craigtaverner committed:
Removed points tests

Because we plan to support points separately.

craigtaverner committed:
Merge branch 'cartesian_doc_values' of github.com:craigtaverner/elasticsearch into cartesian_doc_values

craigtaverner committed:
Merge remote-tracking branch 'origin/master' into cartesian_doc_values

craigtaverner committed:
Revert trivial cleanup as unrelated to PR

craigtaverner committed:
Capitalize CARTESIAN for symmetry with GEO

craigtaverner commented on x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/common/stats/EnumCounters.java:
```diff
@@ -118,11 +118,9 @@ public String toString() {
         boolean first = true;
         for (E e : enums) {
+            if (first == false) buf.append(", ");
+            first = false;
```
https://github.com/elastic/elasticsearch/pull/88505

craigtaverner committed:
Simplify dateline in ShapeDocValuesQuery

craigtaverner committed:
No need for specific generic here

craigtaverner committed:
ShapeIndexer javadocs

craigtaverner committed:
Removed unused class

I originally pulled in some of the other generic refactoring from the
prototype, realised I did not need it and reverted the classes that
extend this class, but forgot to remove this class itself!

craigtaverner committed:
Added tests for setting doc_values in ShapeFieldM.

craigtaverner committed:
Removed older TODO lines

No longer relevant

craigtaverner committed:
Reduced duplication between geo and cartesian test

There were 16 tests shared between the geo and cartesian tests.
These are moved into a base class. The reordering of dependencies
required that we move 3 common dateline tests into a utility class.

iverase reviewed (approved):
LGTM
